### ISSUE TITLE
How to auto convert Module constant Variable between cpu and cuda?

### ISSUE THREAD

**Issue by dianyancao** [pytorch/pytorch#2641]
```
from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.autograd import Variable, Function
from torch.nn import Parameter
from torch.nn.modules.module import Module

# Training settings
parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                    help='input batch size for training (default: 64)')
parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                    help='input batch size for testing (default: 1000)')
parser.add_argument('--epochs', type=int, default=1, metavar='N',
                    help='number of epochs to train (default: 10)')
parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
                    help='learning rate (default: 0.01)')
parser.add_argument('--momentum', type=float, default=0.5, metavar='M',
                    help='SGD momentum (default: 0.5)')
parser.add_argument('--cuda', type=bool, default=True,
                    help='Whether to use CUDA for training')
parser.add_argument('--seed', type=int, default=1, metavar='S',
                    help='random seed (default: 1)')
parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                    help='how many batches to wait before logging training status')
args = parser.parse_args()

import numpy as np
np.random.seed(args.seed)
torch.manual_seed(args.seed)
if args.cuda:
    torch.cuda.manual_seed(args.seed)


kwargs = {'num_workers': 1, 'pin_memory': True} if args.cuda else {}
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=True, download=True,
                   transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ])),
    batch_size=args.batch_size, shuffle=False, **kwargs)
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=False, transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ])),
    batch_size=args.batch_size, shuffle=False, **kwargs)

class ERelu(Module):

    def __init__(self, size):
        super(ERelu, self).__init__()
        self.weight = Parameter(torch.zeros(size))
        self.constant = torch.zeros(1)
        self.constant = Variable(data=self.constant, requires_grad=False)

    def forward(self, input):
        if len(input.size()) == 2:
            input_reshape = input
            t_reshape = self.weight.view(1,-1)
        elif len(input.size()) == 4:
            input_reshape = input
            t_reshape = self.weight.view(1,-1,1,1)
        else:
            raise NotImplementedError()
        max = torch.max(input_reshape - t_reshape, self.constant)
        return (max + t_reshape).view(input.size())
    def cuda(self, device_id=None):
        """Moves all model parameters and buffers to the GPU.

        Arguments:
            device_id (int, optional): if specified, all parameters will be
                copied to that device
        """
        self.constant = self.constant.cuda()
        return self._apply(lambda t: t.cuda(device_id))

    def cpu(self):
        """Moves all model parameters and buffers to the CPU."""
        self.constant = self.constant.cpu()
        return self._apply(lambda t: t.cpu())

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.erelu1 = ERelu(10)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.erelu2 = ERelu(20)
        self.fc1 = nn.Linear(320, 50)
        self.erelu3 = ERelu(50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = self.erelu1((F.max_pool2d(self.conv1(x), 2)))
        x = self.erelu2((F.max_pool2d(self.conv2(x), 2)))
        x = x.view(-1, 320)
        x = self.erelu3((self.fc1(x)))
        x = self.fc2(x)
        return F.log_softmax(x)

model = Net()
if args.cuda:
    model.cuda()

optimizer = optim.SGD(model.parameters(), lr=args.lr, momentum=args.momentum)

def train(epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        if args.cuda:
            data, target = data.cuda(), target.cuda()
        data, target = Variable(data), Variable(target)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % args.log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.2f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.data[0]))

def test():
    model.eval()
    test_loss = 0
    correct = 0
    for data, target in test_loader:
        if args.cuda:
            data, target = data.cuda(), target.cuda()
        data, target = Variable(data, volatile=True), Variable(target)
        output = model(data)
        test_loss += F.nll_loss(output, target, size_average=False).data[0] # sum up batch loss
        #pred = output.data.max(1, keepdim=True)[1] # get the index of the max log-probability
        pred = output.data.max(1)[1]
        correct += pred.eq(target.data.view_as(pred)).cpu().sum()

    test_loss /= len(test_loader.dataset)
    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))


for epoch in range(1, args.epochs + 1):
    train(epoch)
    test()
```
Two questions:
The code run with cuda,I write an ERelu Module to replace Relu activation funtion
but the ERelu.cuda method is not called when I call model.cuda(),how to get work with it?

And I want to fixed all input the same when I run the code next time.
I fix np.random seed, torch cpu cuda random seed, and use shuffle=False for DataLoader,but the test result is different when I run the code next time with cuda,but normal on cpu,is there something to be set?

**Comment by dianyancao** [comment]
'self.register_buffer('constant', Variable(torch.zeros(1),requires_grad=False))'
I register constant to buffer,now it can auto switch between cpu and cuda.

**Comment by soumith** [comment]
`ERelu.cuda` is not called because `.cuda()` just recurses over all it's children and looks for `Parameter` objects, and calls `.cuda()` on them.
Here's the relevant bits:

https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L123-L139

**Comment by apaszke** [comment]
Use `self.register_buffer` for automatic device management, and re-wrap it in a Variable in `forward`. That's the intended usage

**Comment by dianyancao** [comment]
Thanks,I open a new issue for the second question.
